### ISSUE TITLE
How to use the partitioning strategy with the http client?

### ISSUE THREAD

**Issue by gabrielgiussi** [twitter/finagle#953]
**Is your feature request related to a problem? Please describe.**
I want to perform load balancing based on the request to g et affinity with servers that have a hot cache for the data each request will access.
Currently this can't be done in the Load Balancer layer because the pick method doesn't have access to the request.

**Describe the solution you'd like**
I think having a partitioning aware client could be the solution, but I wonder why the service `ConsistentHashPartitioningService` is private, which makes it impossible to reuse with an Http Client.
Even if that class wasn't private, I'm not sure how the wiring should be done, it seems I need to use the Stack API.

**Describe alternatives you've considered**
Using a ServiceFactory to instantiate a service per partition (since it performs LB only once), but this makes using the client much harder since now I have to maintaining this mapping of key to service. 
Besides that, I don't have a way to achieve the affinity, aka telling that service instance which is the subset of nodes that should use to load balance.

**Comment by xin301x** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by gabrielgiussi** [comment]
BTW the `PartitioningService` class says
> At request time it examines the key(s) present in the request and determines the partition(s) it belongs to. It then modifies the original com.twitter.finagle.loadbalancer.LoadBalancerFactory.Dest param and therefore narrows it down to the matching partition(s)

I couldn't figure it out in which part of the code is doing that, in other words, how the `PartitioningService` ends up affecting the set of `Dest` from which the LoadBalancer picks one.

**Comment by gabrielgiussi** [comment]
I've created this module

```scala
package com.twitter.finagle

import com.twitter.finagle.http.{Request, Response}
import com.twitter.finagle.partitioning.{ConsistentHashPartitioningService, PartitioningService}
import com.twitter.finagle.partitioning.param.{KeyHasher, NumReps}
import com.twitter.util.Future
import sun.nio.cs.UTF_8

object HttpPartitioningModule {

  val role = Stack.Role("HttpPartitioning")

  val module: Stackable[ServiceFactory[Request, Response]] = new ConsistentHashPartitioningService.Module[Request,Response,String] {

    override def newConsistentHashPartitioningService(underlying: Stack[ServiceFactory[Request, Response]],
                                                      params: Stack.Params): ConsistentHashPartitioningService[Request, Response, String] = {

      val KeyHasher(hasher) = params[KeyHasher]
      val NumReps(numReps) = params[NumReps]

      new ConsistentHashPartitioningService[Request,Response,String](underlying,params,hasher,numReps) {

        override protected def getKeyBytes(key: String): Array[Byte] = key.getBytes(UTF_8.INSTANCE) 

        override protected def getPartitionKeys(request: Request): Iterable[String] = {
          // TODO assume the header is always present for now
          request.headerMap.get("request-key").toList
        }

        override protected def createPartitionRequestForKeys(original: Request, keys: Seq[String]): Request = {
          // TODO We are not creating multiple requests from one so...
          original
        }

        override protected def mergeResponses(originalReq: Request, results: PartitioningService.PartitionedResults[Request, Response]): Response = {
          // TODO I'm not doing N requests so this should be enough
          results.successes.toList match {
            case (_,res)::_ => res
            case _ => throw results.failures.toList.head._2
          }
        }

        override protected def noPartitionInformationHandler(req: Request): Future[Nothing] = Future.???
      }

    }

    override def role: Stack.Role = HttpPartitioningModule.role

    override def description: String = ""
  }

}

```


And added that module always to the Http Client

```scala
  object Client {
    private val stack: Stack[ServiceFactory[Request, Response]] =
      StackClient.newStack
        .insertBefore(StackClient.Role.prepConn, ClientDtabContextFilter.module)
        .insertBefore(StackClient.Role.prepConn, ClientContextFilter.module)
        // We insert the ClientNackFilter close to the bottom of the stack to
        // eagerly transform the HTTP nack representation to a `Failure`.
        .insertBefore(StackClient.Role.prepConn, ClientNackFilter.module)
        // We add a DelayedRelease module at the bottom of the stack to ensure
        // that the pooling levels above don't discard an active session.
        .replace(StackClient.Role.prepConn, DelayedRelease.module(StackClient.Role.prepConn))
        // Since NackAdmissionFilter should operate on all requests sent over
        // the wire including retries, it must be below `Retries`. Since it
        // aggregates the status of the entire cluster, it must be above
        // `LoadBalancerFactory` (not part of the endpoint stack).
        .replace(
          StackClient.Role.prepFactory,
          NackAdmissionFilter.module[http.Request, http.Response]
        )
        // Ensure that FactoryToService doesn't release the connection to the layers
        // below when the response body hasn't been fully consumed.
        .replace(
          StackClient.Role.requestDraining,
          DelayedRelease.module(StackClient.Role.requestDraining))
        .replace(TraceInitializerFilter.role, new HttpClientTraceInitializer[Request, Response])
        .prepend(http.TlsFilter.module)
        // Because the payload filter also traces the sizes, it's important that we do so
        // after the tracing context is initialized.
        .insertAfter(
          TraceInitializerFilter.role,
          PayloadSizeFilter.module(PayloadSizeFilter.clientTraceKeyPrefix)
        )
        // The pooling strategy depends on whether we're using H2, and if so, what variant.
        // We use a custom module to isolate the selection logic.
        .replace(DefaultPool.Role, HttpPool)
        .prepend(
          new Stack.NoOpModule(http.filter.StatsFilter.role, http.filter.StatsFilter.description)
        )
        .insertAfter(http.filter.StatsFilter.role, StreamingStatsFilter.module)
        // THIS IS ADDED
        .insertAfter(
          BindingFactory.role,
          HttpPartitioningModule.module
        ) 
}
```

With this changes, the request doesn't seem to be passing through the LoadBalancer as it used to be (I've added a breakpoint to `RoundRobinBalancer#Distributor#pick` and it is not passing) although the request is being sent and a status code 200 is returned.

The client code:
```scala
object Client extends App {

  import com.twitter.finagle.{Service, ServiceFactory}
  import com.twitter.finagle.Http
  import com.twitter.finagle.http.{Request, Response}
  import com.twitter.util.{Await, Future}

  val req = Request("/")
  req.headerMap.add("request-key","abc")

  Await.result(Http.client
    .withLoadBalancer(Balancers.roundRobin(2))
    .newService("localhost:8080")
    .apply(req).liftToTry)
    .onSuccess(x => println(x.status.code))
    .onFailure(_.printStackTrace())

}
```

Note: I'm working on top of tag 20.8.1.

**Comment by gabrielgiussi** [comment]
I was able to use the `ConsistentHashPartitioningService` with the Http client and it is working as expected, the `HashRingNodeManager` returns one Service from the ring and this only sees one Host, so whatever load balancer strategy was configured it will always route to that node.

However, I want to also add a new behavior which is fallback to the configured load balancing strategy when the header that includes the partition key is not present.
The `ConsistentHashPartitioningService` doesn't allow such fallback, since we can only return a failed feature when the partition key is not present, overriding `noPartitionInformationHandler`.
The approach I've implemented is to return a Service wrapper that checks for the presence of the header, in which case it forwards to the partitioning service, otherwise **it materializes a ServiceFactory from the stack and forwards to it**.
This last part is the one I don't understand if it is semantically correct and/or safe in terms of resource utilization, but it achieves the intended behavior.

Here is the module 
```scala
import com.twitter.finagle
import com.twitter.finagle.Stack.{Module0, Params}
import com.twitter.finagle.http.{Request, Response}
import com.twitter.finagle.loadbalancer.LoadBalancerFactory
import com.twitter.finagle.partitioning.{ConsistentHashPartitioningService, PartitioningService}
import com.twitter.finagle.partitioning.param.{KeyHasher, NumReps}
import com.twitter.util.Future
import sun.nio.cs.UTF_8

object AffinityHttpClientModule {

  val role = Stack.Role("HttpPartitioning")

  val module = new Stack.Module[ServiceFactory[Request,Response]] {

    val parameters = Seq(
      implicitly[Stack.Param[LoadBalancerFactory.Dest]],
      implicitly[Stack.Param[finagle.param.Stats]]
    )

    def newConsistentHashPartitioningService(underlying: Stack[ServiceFactory[Request, Response]],
                                             params: Params
                                            ): ConsistentHashPartitioningService[Request, Response, String] = {
      val KeyHasher(hasher) = params[KeyHasher]
      val NumReps(numReps) = params[NumReps]

      new ConsistentHashPartitioningService[Request,Response,String](underlying,params,hasher,numReps) {

        override protected def getKeyBytes(key: String): Array[Byte] = key.getBytes(UTF_8.INSTANCE)

        override protected def getPartitionKeys(request: Request): Iterable[String] = {
          request.headerMap.get("request-key").toList
        }

        override protected def createPartitionRequestForKeys(original: Request, keys: Seq[String]): Request = {
          // TODO We are not creating multiple requests from one so...
          original
        }

        override protected def mergeResponses(originalReq: Request, results: PartitioningService.PartitionedResults[Request, Response]): Response = {
          // TODO I'm not doing N requests so this should be enough
          results.successes.toList match {
            case (_,res)::_ => res
            case _ => throw results.failures.toList.head._2
          }
        }

        override protected def noPartitionInformationHandler(req: Request): Future[Nothing] = {
          Future.exception(new Exception())
        }
      }
    }

    final override def make(
                             params: Params,
                             next: Stack[ServiceFactory[Request, Response]]
                           ): Stack[ServiceFactory[Request, Response]] = {
      val partitioningService: Service[Request, Response] = newConsistentHashPartitioningService(next, params)

      // TODO validate if it is ok to materialize the stack here
      val serviceFactory = next.make(params)

      Stack.leaf(role, ServiceFactory.const(new Service[Request,Response]{
        override def apply(request: Request): Future[Response] = {
          partitioningService(request)
          // if the request key is set choose the host based on the consistent hash of the key
          if (request.headerMap.contains("request-key"))
            partitioningService(request)
          else {
            // otherwise, load balance using the configured load balancer
            serviceFactory.apply().flatMap(_.apply(request))
          }

        }
      }))
    }

    override def role: Stack.Role = AffinityHttpClientModule.role

    override def description: String = ""
  }

}
```

And then I do a insertAfter in the http client stack
```scala
  def client: Http.Client = Client().withStack(_.insertAfter(
    BindingFactory.role,
    AffinityHttpClientModule.module
  ))
```
